src/GraphQL/Instrumentation/GraphQLTelemetryProvider.cs

```diff
@@ -0,0 +1,140 @@
+#if NET5_0_OR_GREATER
+
+using System.Diagnostics;
+using GraphQL.DI;
+using GraphQL.Execution;
+using GraphQL.Types;
+using GraphQL.Validation;
+using GraphQLParser.AST;
+
+namespace GraphQL.Instrumentation;
+
+/// <summary>
+/// Provides telemetry through the <see cref="Activity">System.Diagnostics.Activity API</see> as an <see cref="IConfigureExecution"/> implementation.
+/// Derive from this class to add additional telemetry.
+/// <para>
+/// To use a derived class, call <see cref="GraphQLBuilderExtensions.ConfigureExecution{TConfigureExecution}(IGraphQLBuilder)"/> with the type of your derived class.
+/// </para>
+/// </summary>
+public class GraphQLTelemetryProvider : IConfigureExecution
+{
+    private static readonly ActivitySource _activitySource;
+    private readonly GraphQLTelemetryOptions _telemetryOptions;
+    private const string ACTIVITY_OPERATION_NAME = "graphql";
+
+    /// <summary>
+    /// Gets the source name used for telemetry.
+    /// </summary>
+    public static string SourceName => "GraphQL";
+
+    static GraphQLTelemetryProvider()
+    {
+        _activitySource = new ActivitySource(SourceName, typeof(GraphQLTelemetryProvider).Assembly.GetNuGetVersion());
+    }
+
+    /// <summary>
+    /// Initializes a new instance with the specified options.
+    /// </summary>
+    public GraphQLTelemetryProvider(GraphQLTelemetryOptions options)
+    {
+        _telemetryOptions = options;
+    }
+
+    /// <inheritdoc/>
+    public virtual float SortOrder => GraphQLBuilderExtensions.SORT_ORDER_CONFIGURATION;
+
+    /// <inheritdoc/>
+    public virtual async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        // start the Activity
+        using var activity = await StartActivityAsync(options).ConfigureAwait(false);
+
+        // if no event listeners, do not record any telemetry
+        if (activity == null)
+            return await next(options).ConfigureAwait(false);
+
+        // record the requested operation name and optionally the GraphQL document
+        await SetInitialTagsAsync(activity, options).ConfigureAwait(false);
+
+        // record the operation type and the operation name (which may be specified within the
+        // document even if not specified in the request)
+        options.Listeners.Add(new TelemetryListener(this, activity, options));
+
+        // execute the request
+        var result = await next(options).ConfigureAwait(false);
+
+        // record the status
+        await SetResultTagsAsync(activity, options, result).ConfigureAwait(false);
+
+        // return the result
+        return result;
+    }
+
+    /// <summary>
+    /// Creates an <see cref="Activity"/> for the specified <see cref="ExecutionOptions"/> and starts it.
+    /// </summary>
+    protected virtual ValueTask<Activity?> StartActivityAsync(ExecutionOptions options)
```

Comment:
`virtual` here seems useless without ability for the child class to access `_activitySource`.

Reply:
Maybe the ActivitySource should be protected.  Maybe even get/set.  Or the user can just create their own ActivitySource, even if it has the same name.  Or maybe the whole class is sealed.  Let me know what you think.  I'd probably just make ActivitySource protected.

Reply:
protected is the best IMO

> Or the user can just create their own ActivitySource, even if it has the same name.

no, non-friendly way

Reply:
Done; added `protected static ActivitySource ActivitySource { get; }`.

Should it be `get; set;` ?  Should it be `public` instead of `protected`?

Reply:
protected + get-only